{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Assigment 3\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       ":fetch"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "require 'rest-client'\n",
    "\n",
    "def fetch (url, headers = {accept: \"*/*\"}, user = \"\", pass=\"\")\n",
    "    response = RestClient::Request.execute({\n",
    "      method: :get,\n",
    "      url: url.to_s,\n",
    "      user: user,\n",
    "      password: pass,\n",
    "      headers: headers})\n",
    "    \n",
    "    return response\n",
    "  \n",
    "    rescue RestClient::ExceptionWithResponse => e\n",
    "    $stderr.puts e.inspect\n",
    "    response = false\n",
    "    return response\n",
    "  \n",
    "    rescue RestClient::Exception => e\n",
    "    $stderr.puts e.inspect\n",
    "    response = false\n",
    "    return response\n",
    "  \n",
    "    rescue Exception => e\n",
    "    $stderr.puts e.inspect\n",
    "    response = false\n",
    "    return response  \n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "require 'bio'\n",
    "response = fetch('http://www.ebi.ac.uk/Tools/dbfetch/dbfetch?db=ensemblgenomesgene&format=embl&id=At3g54340')\n",
    "if response\n",
    "  record = response.body\n",
    "  sequence = Bio::EMBL.new(record).to_biosequence\n",
    "  #$contains = 0\n",
    "  sequence.features.each do |feature|\n",
    "    if feature.feature == \"exon\"\n",
    "      strand = feature.position.match(/complement/)\n",
    "      exon_region = feature.position.match(/\\(([^\\(]+)\\)/)[1]\n",
    "      exon = Bio::Feature.new('exon',exon_region)\n",
    "      exon.append(Bio::Feature::Qualifier.new('gene', 'At3g54340'))\n",
    "      feature.qualifiers.each do |qualifier|\n",
    "        exon_ID = qualifier.value.match(/(exon_id=((.*?)exon\\d))/)\n",
    "        puts exon_ID\n",
    "        exon.append(Bio::Feature::Qualifier.new('exon_ID', exon_ID[2]))\n",
    "        if not strand.nil?\n",
    "          exon.append(Bio::Feature::Qualifier.new('strand', '-'))\n",
    "          exon_region_limit1 = exon_region.split('..')[0]\n",
    "          exon_region_limit2 = exon_region.split('..')[1]\n",
    "          exon_sequence = sequence.subseq(exon_region_limit1.to_i, exon_region_limit2.to_i).complement\n",
    "          exon.append(Bio::Feature::Qualifier.new('sequence', exon_sequence))\n",
    "          exon.qualifiers.each do |x|\n",
    "            if x.qualifier == \"sequence\" and not x.value.gsub(/(cttctt)+/).map{Regexp.last_match.begin(0)}.empty?\n",
    "              match_positions = x.value.gsub(/(cttctt)+/).map{Regexp.last_match.begin(0)}\n",
    "              match_begin = match_positions.map(&:succ)\n",
    "              match_end = x.value.gsub(/(cttctt)+/).map{Regexp.last_match.end(0)}\n",
    "              i = 0\n",
    "              all_match = Array.new\n",
    "              while i < match_begin.length\n",
    "                match = Bio::Feature.new('match',match_begin[i]..match_end[i])\n",
    "                match.append(Bio::Feature::Qualifier.new('sequid', 'some_gene'))\n",
    "                match.append(Bio::Feature::Qualifier.new('source', 'BioRuby'))\n",
    "                match.append(Bio::Feature::Qualifier.new('type', 'exon'))                \n",
    "                match.append(Bio::Feature::Qualifier.new('score', '.'))\n",
    "                match.append(Bio::Feature::Qualifier.new('phase', '.'))\n",
    "                match.append(Bio::Feature::Qualifier.new('attribute1', 'repetitive_motif=CTTCTT'))\n",
    "                match.append(Bio::Feature::Qualifier.new('attribute2', exon_ID[2]))\n",
    "                all_match << match\n",
    "                i = i +1\n",
    "                #$contains += 1\n",
    "              end\n",
    "              #write_output(all_match)\n",
    "            end \n",
    "          end\n",
    "        else\n",
    "          exon.append(Bio::Feature::Qualifier.new('strand', '+'))\n",
    "          exon_region_limit1 = exon_region.split('..')[0]\n",
    "          exon_region_limit2 = exon_region.split('..')[1]\n",
    "          exon_sequence = sequence.subseq(exon_region_limit1.to_i, exon_region_limit2.to_i)\n",
    "          exon.append(Bio::Feature::Qualifier.new('sequence', exon_sequence))\n",
    "          exon.qualifiers.each do |x|\n",
    "            if x.qualifier == \"sequence\" and not x.value.gsub(/(cttctt)+/).map{Regexp.last_match.begin(0)}.empty?\n",
    "              match_positions = x.value.gsub(/(cttctt)+/).map{Regexp.last_match.begin(0)}\n",
    "              match_begin = match_positions.map(&:succ)\n",
    "              match_end = x.value.gsub(/(cttctt)+/).map{Regexp.last_match.end(0)}\n",
    "              i = 0\n",
    "              all_match = Array.new\n",
    "              while i < match_begin.length\n",
    "                match = Bio::Feature.new('match',match_begin[i]..match_end[i])\n",
    "                match.append(Bio::Feature::Qualifier.new('sequid', 'some_gene'))\n",
    "                match.append(Bio::Feature::Qualifier.new('source', 'BioRuby'))\n",
    "                match.append(Bio::Feature::Qualifier.new('type', 'exon'))               \n",
    "                match.append(Bio::Feature::Qualifier.new('score', '.'))\n",
    "                match.append(Bio::Feature::Qualifier.new('phase', '.'))\n",
    "                match.append(Bio::Feature::Qualifier.new('attribute1', 'repetitive_motif=CTTCTT'))\n",
    "                match.append(Bio::Feature::Qualifier.new('attribute2', exon_ID[2]))\n",
    "                all_match << match\n",
    "                i = i +1 \n",
    "                #$contains += 1\n",
    "              end\n",
    "              #write_output1(all_match)\n",
    "            end \n",
    "          end\n",
    "        end\n",
    "      end   \n",
    "    end\n",
    "  end\n",
    "  #if $contains == 0\n",
    "  #write_no_repetition(some_gene)\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "a = \"heeelloheelloohello\".gsub(/e(e)+/).map{Regexp.last_match.end(0)}\n",
    "#b = a.map(&:succ)\n",
    "a.class\n",
    "for i in a \n",
    "  puts i\n",
    "end\n",
    "a.length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "    case \n",
    "      when feature.feature == \"gene\"\n",
    "      feature.qualifiers.each do |x|\n",
    "        if x.qualifier == \"note\"\n",
    "          source = x.value.match(/Source:(.*?);/)[1]\n",
    "          if not source[1].nil?\n",
    "            source = Bio::Feature::Qualifier.new('source', source)\n",
    "          else\n",
    "            source = Bio::Feature::Qualifier.new('source', 'unknown')\n",
    "          end\n",
    "        end\n",
    "      end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       ":search_exons"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "require 'bio'\n",
    "def search_exons(some_gene)\n",
    "    response = fetch(\"http://www.ebi.ac.uk/Tools/dbfetch/dbfetch?db=ensemblgenomesgene&format=embl&id=#{some_gene}\")\n",
    "    if response\n",
    "      record = response.body\n",
    "      sequence = Bio::EMBL.new(record).to_biosequence\n",
    "      $contains = 0\n",
    "      sequence.features.each do |feature|\n",
    "        if feature.feature == \"exon\"\n",
    "          exon_region = feature.position.match(/\\d+\\.\\.\\d+/).to_s\n",
    "          puts feature.position\n",
    "          exon = Bio::Feature.new('exon',exon_region)\n",
    "          exon.append(Bio::Feature::Qualifier.new('gene', some_gene))\n",
    "          feature.qualifiers.each do |qualifier|\n",
    "            exon_ID = qualifier.value.match(/(exon_id=((.*?)exon\\d))/)[2].to_s\n",
    "            agi = exon_ID.split('.')[0]\n",
    "            if agi.to_s == some_gene.upcase\n",
    "              exon.append(Bio::Feature::Qualifier.new('number', exon_ID))\n",
    "              if not feature.position.match(/complement/)\n",
    "                exon.append(Bio::Feature::Qualifier.new('strand', '+'))\n",
    "                exon_region_limit1 = exon_region.split('..')[0]\n",
    "                exon_region_limit2 = exon_region.split('..')[1]\n",
    "                exon_sequence = sequence.subseq(exon_region_limit1.to_i, exon_region_limit2.to_i)\n",
    "                exon.append(Bio::Feature::Qualifier.new('sequence', exon_sequence))\n",
    "                exon.qualifiers.each do |x|\n",
    "                  puts x.qualifier\n",
    "                  if x.qualifier == \"sequence\" and not x.value.gsub(/ctt(ctt)+/).map{Regexp.last_match.begin(0)}.empty?\n",
    "                    match_positions = x.value.gsub(/ctt(ctt)+/).map{Regexp.last_match.begin(0)}\n",
    "                    match_begin = match_positions.map(&:succ)\n",
    "                    match_end = x.value.gsub(/ctt(ctt)+/).map{Regexp.last_match.end(0)}\n",
    "                    i = 0\n",
    "                    all_match = Array.new\n",
    "                    while i < match_begin.length\n",
    "                      match = Bio::Feature.new('match',match_begin[i]..match_end[i])\n",
    "                      match.append(Bio::Feature::Qualifier.new('sequid', some_gene))\n",
    "                      match.append(Bio::Feature::Qualifier.new('strand', '+'))\n",
    "                      match.append(Bio::Feature::Qualifier.new('attribute2', exon_ID))\n",
    "                      all_match << match\n",
    "                      i = i +1\n",
    "                      $contains += 1\n",
    "                      puts $contains\n",
    "                    end\n",
    "                    ##write_gff1(all_match)\n",
    "                  end\n",
    "                end\n",
    "              else\n",
    "                exon.append(Bio::Feature::Qualifier.new('strand', '-'))\n",
    "                exon_region_limit1 = exon_region.split('..')[0]\n",
    "                exon_region_limit2 = exon_region.split('..')[1]\n",
    "                puts sequence.subseq(exon_region_limit1.to_i, exon_region_limit2.to_i)\n",
    "                exon_sequence = sequence.subseq(exon_region_limit1.to_i, exon_region_limit2.to_i).complement\n",
    "                exon.append(Bio::Feature::Qualifier.new('sequence', exon_sequence))\n",
    "                exon.qualifiers.each do |x|\n",
    "                  if x.qualifier == \"sequence\" and not x.value.gsub(/(cttctt)+/).map{Regexp.last_match.begin(0)}.empty?\n",
    "                    match_positions = x.value.gsub(/(cttctt)+/).map{Regexp.last_match.begin(0)}\n",
    "                    match_begin = match_positions.map(&:succ)\n",
    "                    match_end = x.value.gsub(/(cttctt)+/).map{Regexp.last_match.end(0)}\n",
    "                    i = 0\n",
    "                    all_match = Array.new\n",
    "                    while i < match_begin.length\n",
    "                      match = Bio::Feature.new('match',match_begin[i]..match_end[i])\n",
    "                      match.append(Bio::Feature::Qualifier.new('sequid', some_gene))\n",
    "                      match.append(Bio::Feature::Qualifier.new('strand', '-'))\n",
    "                      match.append(Bio::Feature::Qualifier.new('attribute2', exon_ID))\n",
    "                      all_match << match\n",
    "                      i = i +1\n",
    "                      $contains += 1\n",
    "                      puts $contains\n",
    "                      end\n",
    "                    ##write_gff1(all_match)\n",
    "                  end\n",
    "                end\n",
    "              end\n",
    "            else\n",
    "              next\n",
    "            end\n",
    "            end\n",
    "          end\n",
    "        end\n",
    "      end\n",
    "      if $contains == 0\n",
    "        write_no_cttctt(some_gene)\n",
    "        puts $contains\n",
    "        puts \"soy bobo\"\n",
    "      end\n",
    "    end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "F10K1_1:83276..83551\n",
      "F10K1_1:83674..84216\n",
      "2085..2441\n",
      "gene\n",
      "number\n",
      "strand\n",
      "sequence\n",
      "1922..1991\n",
      "gene\n",
      "number\n",
      "strand\n",
      "sequence\n",
      "953..1165\n",
      "gene\n",
      "number\n",
      "strand\n",
      "sequence\n",
      "1..335\n",
      "gene\n",
      "number\n",
      "strand\n",
      "sequence\n",
      "1\n",
      "399..511\n",
      "gene\n",
      "number\n",
      "strand\n",
      "sequence\n",
      "1750..1838\n",
      "gene\n",
      "number\n",
      "strand\n",
      "sequence\n",
      "1255..1390\n",
      "gene\n",
      "number\n",
      "strand\n",
      "sequence\n",
      "2085..2472\n",
      "gene\n",
      "number\n",
      "strand\n",
      "sequence\n",
      "707..808\n",
      "gene\n",
      "number\n",
      "strand\n",
      "sequence\n",
      "83..335\n",
      "gene\n",
      "number\n",
      "strand\n",
      "sequence\n",
      "2\n",
      "695..808\n",
      "gene\n",
      "number\n",
      "strand\n",
      "sequence\n",
      "594..626\n",
      "gene\n",
      "number\n",
      "strand\n",
      "sequence\n",
      "1477..1652\n",
      "gene\n",
      "number\n",
      "strand\n",
      "sequence\n",
      "55..335\n",
      "gene\n",
      "number\n",
      "strand\n",
      "sequence\n",
      "3\n",
      "426..526\n",
      "gene\n",
      "number\n",
      "strand\n",
      "sequence\n",
      "complement(F10K1_1:83807..84216)\n"
     ]
    }
   ],
   "source": [
    "search_exons('AT1g07010')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"84216\""
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = \"83674..84216\"\n",
    "a.split(\"..\")[1]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Ruby 2.5.1",
   "language": "ruby",
   "name": "ruby"
  },
  "language_info": {
   "file_extension": ".rb",
   "mimetype": "application/x-ruby",
   "name": "ruby",
   "version": "2.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
